### PR TITLE
[SDFAB-1002,1028] Implementing Cascading Delete functionality

### PR DESCRIPTION
**Pull request by eroshiva**
This PR implements Cascading Delete functionality inside onos-config component. Cascading delete means that when the intermediate path is being deleted, it also deletes all its children.
Integration test is also a part of that PR.

### CODE REVIEW

File: pkg/controller/proposal/controller.go
Comment (SeanCondon) [inline pkg/controller/proposal/controller.go:824]: I think a better name might be `addDeletedChildren` and the comment should go on the line before

File: pkg/controller/proposal/controller.go
Comment (SeanCondon) [inline pkg/controller/proposal/controller.go:459]: this `addDeletedChildren` should be conditional on a flag from modelPlugin ModelInfo

File: pkg/controller/proposal/controller.go
Comment (SeanCondon) [inline pkg/controller/proposal/controller.go:646]: I think `upd` is a bit odd - suggest `updated`

File: pkg/controller/proposal/controller.go
Comment (SeanCondon) [inline pkg/controller/proposal/controller.go:832]: I did not understand the comment re line 473.
Also consider that a deleted object and a child might already be in the `changeValues` - it should not matter as the map will guarantee uniqueness

File: pkg/controller/proposal/cascading-delete_test.go
Comment (SeanCondon) [inline pkg/controller/proposal/cascading-delete_test.go:113]: I think you should be checking the contents too

File: pkg/controller/proposal/controller.go
Comment (eroshiva) [reply pkg/controller/proposal/controller.go]: Renamed

File: pkg/controller/proposal/controller.go
Comment (eroshiva) [reply pkg/controller/proposal/controller.go]: Changed per suggestion

File: pkg/controller/proposal/controller.go
Comment (eroshiva) [reply pkg/controller/proposal/controller.go]: Discussed in a call - no need to implement a flag so far

File: pkg/controller/proposal/cascading-delete_test.go
Comment (eroshiva) [reply pkg/controller/proposal/cascading-delete_test.go]: Added

File: pkg/controller/proposal/controller.go
Comment (eroshiva) [reply pkg/controller/proposal/controller.go]: Indeed, this comment is very confusing. I completely removed it..

**Comment by eroshiva** [comment]
Pinging @adibrastegarnia and @kuujo for a review as well

**Review by Andrea-Campanella — APPROVED** [review]
Generally LGTM, I need to test this with the autogenerated ui, cc @teone

File: test/config/cascadingdelete.go
Comment (adibrastegarnia) [inline test/config/cascadingdelete.go:20]: You need to change these paths to use gnmi simulator model for testing

**Review by adibrastegarnia — CHANGES_REQUESTED** [review]

File: test/config/cascadingdelete.go
Comment (eroshiva) [reply test/config/cascadingdelete.go]: Adjusted per our discussion

**Review by Andrea-Campanella — COMMENTED** [review]
great work! I'd say we wait to merge this until @charlesmcchan makes the release. It's a known issue, might want to do a point release with this and removing the workaround in UI and ONOS. cc also @daniele-moro

**Comment by teone** [comment]
@Andrea-Campanella I don't think this level changes will have any effect on the UI.
@frouzbeh Just out of caution, if you have spare cycles today, can you try to run it with the sd-fabric models and the latest UI?

**Comment by eroshiva** [comment]
retest this please

**Comment by adibrastegarnia** [comment]
The integration test for cascading delete need to be fixed before we can merge this.

**Comment by SeanCondon** [comment]
> @Andrea-Campanella I don't think this level changes will have any effect on the UI. @frouzbeh Just out of caution, if you have spare cycles today, can you try to run it with the sd-fabric models and the latest UI?

Agreed - this has nothing to do with the UI

**Comment by adibrastegarnia** [comment]
@eroshiva 
FYI: I noticed some of the basic integration tests (specially the ones that delete leaf values) also affected (we have a bunch of them in test folder). If you were seeing the tests were passing perhaps you were deploying a wrong image that doesn't have your changes or something like that.

**Comment by frouzbeh** [comment]
@teone @Andrea-Campanella 
I agree these changes has nothing to do with the gui, but I've tested and I wasn't able  to delete even leaf attributes. 
Previously, when we deleted a top-level model (e.g. switch), we had `deleted: true` for this  path `/switch[switch-id=<id>]`, but not its leaves and attributes. So the gui wouldn't show the model because the top-level model was tagged with deleted. 
With these changes,  I still can see the "deleted" switch in the gui, and I noticed there is no `deleted` tag.

**Comment by Andrea-Campanella** [comment]
> @teone @Andrea-Campanella
> I agree these changes has nothing to do with the gui, but I've tested and I wasn't able to delete even leaf attributes.
> Previously, when we deleted a top-level model (e.g. switch), we had `deleted: true` for this path `/switch[switch-id=<id>]`, but not its leaves and attributes. So the gui wouldn't show the model because the top-level model was tagged with deleted.
> With these changes, I still can see the "deleted" switch in the gui, and I noticed there is no `deleted` tag.

AN update here, @eroshiva found issues in his implementation that he is working out with @adibrastegarnia. please refrain from further testing until notified. thanks for all the effort.

File: pkg/controller/proposal/controller.go
Comment (SeanCondon) [inline pkg/controller/proposal/controller.go:460]: it might be better to make log.Debugf, and set the logging level to activate - then they could be left in permanently

File: pkg/controller/utils/utils.go
Comment (Andrea-Campanella) [inline pkg/controller/utils/utils.go:27]: might want to remove this log as well

File: pkg/controller/proposal/controller.go
Comment (Andrea-Campanella) [inline pkg/controller/proposal/controller.go:463]: I think we agreed of handing this behind a variable, maybe not all targets want a cascading delete to happen. @adibrastegarnia @SeanCondon am I correct ?

**Review by Andrea-Campanella — COMMENTED** [review]
code mostly looks good, 2 comments. Once those are addressed I'd like @ray-milkey and @frouzbeh to test this please. I will wait for their review before any merge.

File: pkg/controller/proposal/controller.go
Comment (adibrastegarnia) [reply pkg/controller/proposal/controller.go:463]: I think cascading delete is something generic. That means when you delete an intermediate node in the config tree, it makes sense the whole subtree gets deleted. If the user doesn't want to delete the whole subtree then the user should just delete leaf values or just the paths that it cares about. I don't think enabling/disabling per target makes sense in general.

**Comment by eroshiva** [comment]
retest this please

File: pkg/controller/utils/utils.go
Comment (eroshiva) [reply pkg/controller/utils/utils.go]: removed

File: pkg/controller/proposal/controller.go
Comment (eroshiva) [reply pkg/controller/proposal/controller.go]: Log was completely removed - no need to keep it

**Review by adibrastegarnia — APPROVED** [review]

**Comment by frouzbeh** [comment]
@eroshiva @Andrea-Campanella 
I just tested this pr with the adaptive gui, and it seems everything works as expected. I was able to delete from list, or unset values. And in the configuration all deleted paths are set with `deleted: true` as expected.

**Pull request merged**
